File: sdks/typescript/src/resources/sessions.ts
import { randomUUID } from "node:crypto";

import type { Command, CommandCreateRequest, Session } from "../generated/types.ts";
import type { Palai } from "../client.ts";
import { callArgs, enc, listPath, type CallOptions, type ListParams, type Page } from "./shared.ts";

// SteerParams / InterruptParams carry the message a steer/interrupt delivers. command_id is the
// command's own idempotency handle (spec §22.4): the SDK mints a stable one per call and lets a
// caller supply their own to dedupe across processes — mirroring the idempotency key in
// responses.create.
export interface SteerParams {
  message: string;
  commandId?: string;
}
export type InterruptParams = SteerParams;

// SessionCreateParams carries the one thing a caller may supply when opening a session: its display
// label. Omitting it is the pre-existing bodyless create, unchanged.
export interface SessionCreateParams {
  name?: string;
}

// SessionCommands is the durable command surface of a session (spec §9.2, §22.4) — E08's steering
// product, used from the SDK for the first time. steer and interrupt are both send_message
// deliveries; the delivery mode is the only difference (queue is the default fire path and not
// exposed here — the plan asks for steer + interrupt).
export class SessionCommands {
  #client: Palai;
  constructor(client: Palai) {
    this.#client = client;
  }

  // steer delivers a message at the next safe boundary WITHOUT interrupting the current step
  // (delivery=steer). Acceptance is durable-queued (202), not applied; the returned Command is the
  // queued handle, or — on a duplicate command_id — the original unchanged (idempotent).
  steer(sessionID: string, params: SteerParams, options: CallOptions = {}): Promise<Command> {
    return this.#send(sessionID, "steer", params, options);
  }

  // interrupt delivers a message that preempts the current step (delivery=interrupt). Same durable,
  // idempotent acceptance as steer.
  interrupt(sessionID: string, params: InterruptParams, options: CallOptions = {}): Promise<Command> {
    return this.#send(sessionID, "interrupt", params, options);
  }

  async #send(sessionID: string, delivery: "steer" | "interrupt", params: SteerParams, options: CallOptions): Promise<Command> {
    const body: CommandCreateRequest = {
      command_id: params.commandId ?? newCommandId(),
      kind: "send_message",
      delivery,
      message: params.message,
    };
    const result = await this.#client.request<Command>("POST", `/v1/sessions/${enc(sessionID)}/commands`, {
      body,
      // command_id carries idempotency server-side, so a network re-send settles exactly one command.
      idempotent: true,
      ...callArgs(options),
    });
    return result.body;
  }
}

// Sessions is the /v1/sessions resource: create, retrieve, list, and the durable command surface.
export class Sessions {
  #client: Palai;
  readonly commands: SessionCommands;

  constructor(client: Palai) {
    this.#client = client;
    this.commands = new SessionCommands(client);
  }

  // create opens a standalone session (201). Creation is cheap and unkeyed — a retried create mints
  // a new session — so it carries no idempotency key. `name` is the optional display label; without
  // one the session's projection derives a label from its first prompt (see `rename`).
  async create(params: SessionCreateParams = {}, options: CallOptions = {}): Promise<Session> {
    const args = callArgs(options);
    const result = await this.#client.request<Session>(
      "POST",
      "/v1/sessions",
      params.name === undefined ? args : { body: { name: params.name }, ...args },
    );
    return result.body;
  }

  // rename sets a session's display label and returns the re-read session, so the caller sees both
  // the label and the `name_source` that now reads "operator".
  //
  // This, and not the create-time name, is the call a session list needs: most sessions are never
  // created through `create` at all — posting a response opens one implicitly — so they reach a list
  // screen with no label to have been given one. Passing an EMPTY string is meaningful: it clears the
  // operator label and returns the row to its derived one.
  async rename(sessionID: string, name: string, options: CallOptions = {}): Promise<Session> {
    const result = await this.#client.request<Session>("PATCH", `/v1/sessions/${enc(sessionID)}`, {
      body: { name },
      // Setting a label to X twice leaves it X, so a connection torn after the server committed is
      // safe to re-send. That is a property of THIS call, not of PATCH, which is why it is declared
      // here rather than in the transport's method table.
      idempotent: true,
      ...callArgs(options),
    });
    return result.body;
  }

  // retrieve reads a session within the verified scope; an unknown or foreign id is a 404
  // (NotFoundError).
  async retrieve(sessionID: string, options: CallOptions = {}): Promise<Session> {
    const result = await this.#client.request<Session>("GET", `/v1/sessions/${enc(sessionID)}`, callArgs(options));
    return result.body;
  }

  // list returns a tenant-scoped page of sessions with the shared cursor + basic filters.
  async list(params: ListParams = {}, options: CallOptions = {}): Promise<Page<Session>> {
    const result = await this.#client.request<Page<Session>>("GET", listPath("/v1/sessions", params), callArgs(options));
    return result.body;
  }
}

function newCommandId(): string {
  return `cmd_${randomUUID()}`;
}

File: sdks/go/sessions.go
package palai

import (
	"context"
	"crypto/rand"
	"encoding/hex"
	"encoding/json"
	"net/http"
	"strconv"
	"time"
)

// Sessions is the /v1/sessions resource group (spec §9.1, §9.2): open a standalone session and
// deliver a durable steer command into it. Mirrors Responses's shape (a struct hanging off *Client,
// each method taking a trailing opts ...CallOption) rather than the TS SDK's separate
// Sessions/SessionCommands split — Steer lives directly on Sessions since this task ships only the
// one delivery mode; a later task can still add a nested Commands group without breaking this one.
type Sessions struct{ client *Client }

// CreateSessionParams carries the one thing a caller may supply when opening a session: its
// display label (E29). Mirrors the TS SDK's SessionCreateParams; the server's SessionWrite decode
// is strict (DisallowUnknownFields), so a field beyond `name` is a 400, not a silently ignored one.
type CreateSessionParams struct {
	Name string `json:"name,omitempty"`
}

// SteerParams carries the message a steer delivers (spec §22.4): the command applies at the next
// safe boundary WITHOUT interrupting the session's current step. CommandID is the command's own
// idempotency handle — the SDK mints a stable one per call and a caller may supply its own to
// dedupe across processes, mirroring the Idempotency-Key pattern Responses.Create uses for creates.
type SteerParams struct {
	Message   string
	CommandID string
}

// commandCreateRequest is the wire body for POST /v1/sessions/{id}/commands (spec §22.4, §9.2).
// It is unexported: Sessions.Steer is the only delivery mode this task ships (kind=send_message,
// delivery=steer) — queue/interrupt and the other command kinds are a later task's surface.
type commandCreateRequest struct {
	CommandID string `json:"command_id"`
	Kind      string `json:"kind"`
	Delivery  string `json:"delivery"`
	Message   string `json:"message"`
}

// Session is a standalone session's projection (spec §9.1). Known fields are typed for ergonomics;
// any field a newer server adds is preserved in Extra and re-emitted on marshal (spec API-009),
// matching every other decoded model in this SDK (see types.go).
type Session struct {
	ID     string                     `json:"id"`
	Object string                     `json:"object"`
	Status string                     `json:"status"`
	Extra  map[string]json.RawMessage `json:"-"`
}

func (s *Session) UnmarshalJSON(b []byte) error {
	type alias Session
	var a alias
	if err := forwardUnmarshal(b, &a, &a.Extra); err != nil {
		return err
	}
	*s = Session(a)
	return nil
}

func (s Session) MarshalJSON() ([]byte, error) {
	type alias Session
	return forwardMarshal(alias(s), s.Extra)
}

// Command is a session's durable command (spec §22.4, §9.2). Acceptance means durably queued, not
// applied: Status starts "queued", and a later terminal state arrives over the session's event
// stream, not by re-fetching this struct.
type Command struct {
	ID              string                     `json:"id"`
	Object          string                     `json:"object"`
	SessionID       string                     `json:"session_id"`
	Kind            string                     `json:"kind"`
	Delivery        string                     `json:"delivery,omitempty"`
	Status          string                     `json:"status"`
	CreatedAt       string                     `json:"created_at,omitempty"`
	AppliedSequence *int                       `json:"applied_sequence,omitempty"`
	Result          map[string]any             `json:"result,omitempty"`
	Extra           map[string]json.RawMessage `json:"-"`
}

func (c *Command) UnmarshalJSON(b []byte) error {
	type alias Command
	var a alias
	if err := forwardUnmarshal(b, &a, &a.Extra); err != nil {
		return err
	}
	*c = Command(a)
	return nil
}

func (c Command) MarshalJSON() ([]byte, error) {
	type alias Command
	return forwardMarshal(alias(c), c.Extra)
}

// Create opens a standalone session (201). Creation is cheap and unkeyed — a retried create mints
// a NEW session, never the same one back — so, unlike Responses.Create, this carries no
// Idempotency-Key: a connection torn after the server committed is left alone rather than retried
// into a duplicate session.
func (s *Sessions) Create(ctx context.Context, p CreateSessionParams, opts ...CallOption) (*Session, error) {
	o := requestOptions{body: p}
	for _, opt := range opts {
		opt(&o)
	}
	var out Session
	if err := s.client.doJSON(ctx, http.MethodPost, "/v1/sessions", o, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// Steer posts a durable send_message/steer command against a session (spec §22.4). It is 202
// (durably queued, not yet applied). CommandID rides in the body as the server's own dedupe key
// (the command table's unique on it) — a duplicate command_id returns the original resource
// unchanged, so a retried steer settles exactly one command; that is why this call, unlike Create,
// defaults to idempotent for the transport's network-failure retry.
func (s *Sessions) Steer(ctx context.Context, sessionID string, p SteerParams, opts ...CallOption) (*Command, error) {
	commandID := p.CommandID
	if commandID == "" {
		commandID = newCommandID()
	}
	body := commandCreateRequest{
		CommandID: commandID,
		Kind:      "send_message",
		Delivery:  "steer",
		Message:   p.Message,
	}
	o := requestOptions{body: body, idempotent: true}
	for _, opt := range opts {
		opt(&o)
	}
	path := "/v1/sessions/" + escapePathSegment(sessionID) + "/commands"
	var out Command
	if err := s.client.doJSON(ctx, http.MethodPost, path, o, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// newCommandID mints a fresh command idempotency handle (spec §22.4), the same entropy source as
// newIdempotencyKey with the command-scoped prefix the TS SDK's `cmd_${randomUUID()}` also uses.
func newCommandID() string {
	var b [16]byte
	if _, err := rand.Read(b[:]); err != nil {
		return "cmd_" + strconv.FormatInt(time.Now().UnixNano(), 36)
	}
	return "cmd_" + hex.EncodeToString(b[:])
}
